clang/test/dpct/dnn/types.cu

```diff
@@ -0,0 +1,73 @@
+// RUN: dpct -in-root %S -out-root %T/types %S/types.cu --cuda-include-path="%cuda-path/include" -- -std=c++14 -x cuda --cuda-host-only
+// RUN: FileCheck --input-file %T/types/types.dp.cpp --match-full-lines %s
+#include <cuda_runtime.h>
+#include <cudnn.h>                                                                  
+
+struct cudnnHandleWrapper {
+  cudnnHandle_t handle;
+  cudnnHandleWrapper() { cudnnCreate(&handle); }    
+  cudnnHandleWrapper(const cudnnHandleWrapper&) = delete;
+  cudnnHandleWrapper(cudnnHandleWrapper&& from)
+    : handle(from.handle)
+      // CHECK:      { // begin
+      // CHECK-NEXT: } // end
+  { // begin
+    from.handle = NULL;
+    from.handle = 0;
+    from.handle = nullptr;
+  } // end
+
+  // CHECK:      ~cudnnHandleWrapper() {
+  // CHECK-NEXT:   if (true) {
+  ~cudnnHandleWrapper() {
+    if (handle) {
+      cudnnDestroy(handle);
+    }
+  }
+};
+
+struct cudnnTensorDescriptorWrapper {
+  cudnnTensorDescriptor_t handle;
+  cudnnTensorDescriptorWrapper() { cudnnCreateTensorDescriptor(&handle); }    
+  cudnnTensorDescriptorWrapper(const cudnnTensorDescriptorWrapper&) = delete;
+  cudnnTensorDescriptorWrapper(cudnnTensorDescriptorWrapper&& from)
+    : handle(from.handle)
+      // CHECK:      { // begin
+      // CHECK-NEXT: } // end
+  { // begin
+    from.handle = NULL;
+    from.handle = 0;
+    from.handle = nullptr;
+  } // end
+
+  // CHECK:      ~cudnnTensorDescriptorWrapper() {
+  // CHECK-NEXT:   if (true) {
+  ~cudnnTensorDescriptorWrapper() {
+    if (handle) {
+      cudnnDestroyTensorDescriptor(handle);
+    }
+  }
+};
+
+struct cudnnConvolutionDescriptorWrapper {
+  cudnnConvolutionDescriptor_t handle;
+  cudnnConvolutionDescriptorWrapper() { cudnnCreateConvolutionDescriptor(&handle); }    
+  cudnnConvolutionDescriptorWrapper(const cudnnConvolutionDescriptorWrapper&) = delete;
+  cudnnConvolutionDescriptorWrapper(cudnnConvolutionDescriptorWrapper&& from)
+    : handle(from.handle)
+      // CHECK:      { // begin
+      // CHECK-NEXT: } // end
+  { // begin
+    from.handle = NULL;
+    from.handle = 0;
+    from.handle = nullptr;
+  } // end
+
+  // CHECK:      ~cudnnConvolutionDescriptorWrapper() {
+  // CHECK-NEXT: if (true) {
```

Comment:
I think use operator= overload and user defined conversion function is a better solution.
Just like:
operator=(std::nullptr_t ptr){
 set the _q to null and clear worksapce
}
and 
operator bool(){
 return q == nullptr;
}